kimchi/src/tests/foreign_field_mul.rs

```diff
@@ -0,0 +1,1409 @@
+        ForeignFieldHelpers,
+    },
+    FieldHelpers,
+};
+
+use mina_poseidon::{
+    constants::PlonkSpongeConstantsKimchi,
+    sponge::{DefaultFqSponge, DefaultFrSponge},
+    FqSponge,
+};
+use num_bigint::RandBigInt;
+use rand::{rngs::StdRng, SeedableRng};
+
+type PallasField = <Pallas as AffineCurve>::BaseField;
+type VestaField = <Vesta as AffineCurve>::BaseField;
+
+type SpongeParams = PlonkSpongeConstantsKimchi;
+type VestaBaseSponge = DefaultFqSponge<VestaParameters, SpongeParams>;
+type VestaScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
+type PallasBaseSponge = DefaultFqSponge<PallasParameters, SpongeParams>;
+type PallasScalarSponge = DefaultFrSponge<Fq, SpongeParams>;
+
+const RNG_SEED: [u8; 32] = [
+    211, 31, 143, 75, 29, 255, 0, 126, 237, 193, 86, 160, 1, 90, 131, 221, 186, 168, 4, 95, 50, 48,
+    89, 29, 13, 250, 215, 172, 130, 24, 164, 162,
+];
+
+// The secp256k1 base field modulus
+fn secp256k1_modulus() -> BigUint {
+    BigUint::from_bytes_be(&secp256k1::constants::FIELD_SIZE)
+}
+
+// Maximum value in the secp256k1 base field
+fn secp256k1_max() -> BigUint {
+    secp256k1_modulus() - BigUint::from(1u32)
+}
+
+// Maximum value whose square fits in secp256k1 base field
+fn secp256k1_sqrt() -> BigUint {
+    secp256k1_max().sqrt()
+}
+
+// Maximum value in the pallas base field
+fn pallas_max() -> BigUint {
+    PallasField::modulus_biguint() - BigUint::from(1u32)
+}
+
+// Maximum value whose square fits in the pallas base field
+fn pallas_sqrt() -> BigUint {
+    pallas_max().sqrt()
+}
+
+// Boilerplate for tests
+fn run_test<G: KimchiCurve, EFqSponge, EFrSponge>(
+    full: bool,
+    external_gates: bool,
+    target_plookups: bool,
+    left_input: &BigUint,
+    right_input: &BigUint,
+    foreign_field_modulus: &BigUint,
+    invalidations: Vec<((usize, usize), G::ScalarField)>,
+) -> (CircuitGateResult<()>, [Vec<G::ScalarField>; COLUMNS])
+where
+    G::BaseField: PrimeField,
+    EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    EFrSponge: FrSponge<G::ScalarField>,
+{
+    // Create foreign field multiplication gates
+    let (mut next_row, mut gates) = CircuitGate::<G::ScalarField>::create_foreign_field_mul(0);
+
+    // Compute multiplication witness
+    let (mut witness, external_checks) =
+        foreign_field_mul::witness::create(left_input, right_input, foreign_field_modulus);
+
+    // Optionally also add external gate checks to circuit
+    if external_gates {
+        // Layout for this test (just an example, circuit designer has complete flexibility, where to put the checks)
+        //      0-1  ForeignFieldMul
+        //      2-3  ForeignFieldAdd (result bound addition)
+        //      4-7  multi-range-check (left multiplicand)
+        //      8-11 multi-range-check (right multiplicand)
+        //     12-15 multi-range-check (product1_lo, product1_hi_0, carry1_lo)
+        //     16-19 multi-range-check (result range check)
```

Comment:
refresh my memory please... what is this range check exactly doing? Checking that result is <2^264 or that result is < fmod? It must be the latter right? copying from the result bound addition output

Reply:
$< 2^{264}$ (it's part of matthew's $< f$ check)